rct_optimizations/test/covariance_utest.cpp

```diff
@@ -0,0 +1,455 @@
+  EXPECT_NEAR(result.covariance(0, 0), result.covariance(1, 1), 1e-5);
+}
+
+TEST_F(CircleFitUnit_ClusteredObservations, FitCircleToClusteredObs)
+{
+  rct_optimizations::CircleFitResult result;
+  EXPECT_NO_THROW(result = rct_optimizations::optimize(problem));
+  printResults(result);
+
+  // assert 3x3 covariance matrix
+  ASSERT_EQ(result.covariance.rows(), 3);
+  ASSERT_EQ(result.covariance.cols(), 3);
+
+  // expect matrix to be symmetric
+  EXPECT_NEAR(result.covariance(0, 1), result.covariance(1, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(0, 2), result.covariance(2, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(1, 2), result.covariance(2, 1), std::numeric_limits<double>::epsilon());
+
+  // expect diagonal elements to be positive
+  EXPECT_GE(result.covariance(0, 0), 0.0);
+  EXPECT_GE(result.covariance(1, 1), 0.0);
+  EXPECT_GE(result.covariance(2, 2), 0.0);
+
+  // std dev of Y-coord should be greater than std dev of X-coord
+  EXPECT_GT(result.covariance(1, 1), result.covariance(0, 0));
+}
+
+TEST_F(CircleFitUnit_ClusteredObservationsPerturbed, FitCircleToClusteredObsPerturbed)
+{
+  rct_optimizations::CircleFitResult result;
+  EXPECT_NO_THROW(result = rct_optimizations::optimize(problem));
+  printResults(result);
+
+  // assert 3x3 covariance matrix
+  ASSERT_EQ(result.covariance.rows(), 3);
+  ASSERT_EQ(result.covariance.cols(), 3);
+
+  // expect matrix to be symmetric
+  EXPECT_NEAR(result.covariance(0, 1), result.covariance(1, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(0, 2), result.covariance(2, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(1, 2), result.covariance(2, 1), std::numeric_limits<double>::epsilon());
+
+  // expect diagonal elements to be positive
+  EXPECT_GE(result.covariance(0, 0), 0.0);
+  EXPECT_GE(result.covariance(1, 1), 0.0);
+  EXPECT_GE(result.covariance(2, 2), 0.0);
+
+  // std dev of Y-coord should be greater than std dev of X-coord
+  EXPECT_GT(result.covariance(1, 1), result.covariance(0, 0));
+}
+
+TEST_F(CircleFitUnit_ParallelLines, FitCircleToParallelLines)
+{
+  rct_optimizations::CircleFitResult result;
+  EXPECT_NO_THROW(result = rct_optimizations::optimize(problem));
+  printResults(result);
+
+  // assert 3x3 covariance matrix
+  ASSERT_EQ(result.covariance.rows(), 3);
+  ASSERT_EQ(result.covariance.cols(), 3);
+
+  // expect matrix to be symmetric
+  EXPECT_NEAR(result.covariance(0, 1), result.covariance(1, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(0, 2), result.covariance(2, 0), std::numeric_limits<double>::epsilon());
+  EXPECT_NEAR(result.covariance(1, 2), result.covariance(2, 1), std::numeric_limits<double>::epsilon());
+
+  // expect diagonal elements to be positive
+  EXPECT_GE(result.covariance(0, 0), 0.0);
+  EXPECT_GE(result.covariance(1, 1), 0.0);
+  EXPECT_GE(result.covariance(2, 2), 0.0);
+}
```

Comment:
I was expecting that it would, but due to the cost model for this optimization it ends up converging with low variance/covariance and somewhat higher than average cost. Probably should just remove it since it doesn't contribute as much as the other tests.